### PR TITLE
Add encoder_freeze method to SegmentationModel

### PR DESCRIPTION
**Pull request by ricber**
### Summary

This PR introduces a new method, `encoder_freeze`, in the `SegmentationModel` base class. The method allows users to freeze all encoder parameters and prevent updates to normalization layer statistics (`BatchNorm`/`InstanceNorm`) during training. Related to #1197 

### Motivation

When fine-tuning segmentation models, it is common to freeze the encoder (pre-trained backbone) while training only the decoder. This reduces training time and avoids overfitting on small datasets. Currently, users must manually set `requires_grad=False` and handle `BatchN### Summary

This PR introduces a new method, `encoder_freeze`, in the `SegmentationModel` base class. The method allows users to freeze all encoder parameters and prevent updates to normalization layer statistics (`BatchNorm`/`InstanceNorm`) during training. Related to #1197 

### Motivation

When fine-tuning segmentation models, it is common to freeze the encoder (pre-trained backbone) while training only the decoder. This reduces training time and avoids overfitting on small datasets. Currently, users must manually set `requires_grad=False` and handle normalization layers, which is error-prone.

`encoder_freeze` automates this process:

- Freezes all encoder weights.

- Puts BatchNorm/InstanceNorm layers in `eval()` mode to stop updates to `running_mean` and `running_var`.

**Important**: Calling `model.train()` after `encoder_freeze` will re-enable updates to the encoder’s normalization layers’ running statistics. Therefore, this method is intended to be called again after each `model.train()` invocation.

### Testing

I ran a small training loop on dummy data using a `Unet(resnet18)` model and confirmed that the encoder parameters remained frozen. I also verified that the batch normalization running statistics stayed fixed.

### Future Work

- If these changes are satisfactory, I can update the documentation accordingly.
- Consider partially freezing only some layers of the encoder.orm` layers, which is error-prone.

`encoder_freeze` automates this process:

- Freezes all encoder weights.

- Puts BatchNorm/InstanceNorm layers in `eval()` mode to stop updates to `running_mean` and `running_var`.

**Important**: Calling `model.train()` after `encoder_freeze` will re-enable updates to the encoder’s normalization layers’ running statistics. Therefore, this method is intended to be called again after each `model.train()` invocation.

### Testing

I ran a small training loop on dummy data using a `Unet(resnet18)` model and confirmed that the encoder parameters remained frozen. I also verified that the batch normalization running statistics stayed fixed.

### Future Work

- If these changes are satisfactory, I can update the documentation accordingly.
- Consider partially freezing only some layers of the encoder.

### CODE REVIEW

File: segmentation_models_pytorch/base/model.py [hunk 1]
```diff
@@ -137,3 +137,29 @@ def load_state_dict(self, state_dict, **kwargs):
                 warnings.warn(text, stacklevel=-1)
 
         return super().load_state_dict(state_dict, **kwargs)
+
+    def encoder_freeze(self):
```
Comment (qubvel) [inline segmentation_models_pytorch/base/model.py:141]: let's name it `freeze_encoder`, what do you think?

File: segmentation_models_pytorch/base/model.py [hunk 2]
```diff
@@ -137,3 +137,29 @@ def load_state_dict(self, state_dict, **kwargs):
...
+        preventing them from being updated during backpropagation. In addition,
+        it switches BatchNorm and InstanceNorm layers into evaluation mode
+        (``.eval()``), which stops updates to their running statistics
+        (``running_mean`` and ``running_var``) during training.
+
+        **Important:** If you call :meth:`model.train()` after
+        :meth:`encoder_freeze`, the encoder’s BatchNorm/InstanceNorm layers
+        will be put back into training mode, and their running statistics
+        will be updated again. To re-freeze them after switching modes,
+        call :meth:`encoder_freeze` again.
```
Comment (qubvel) [inline segmentation_models_pytorch/base/model.py:155]: To make the user's experience better, we can override `train` method for the model similar to the [original](https://github.com/pytorch/pytorch/blob/ba56102387ef21a3b04b357e5b183d48f0afefc7/torch/nn/modules/module.py#L2867) one

```py
    def train(self, mode: bool = True) -> Self:
        if not isinstance(mode, bool):
            raise ValueError("training mode is expected to be boolean")
        self.training = mode
        for name, module in self.named_children():
            # skip encoder encoder if it is frozen
            if self._is_encoder_frozen and name == "encoder":
                continue
            module.train(mode)
        return self
```

File: segmentation_models_pytorch/base/model.py [hunk 3]
```diff
@@ -137,3 +137,29 @@ def load_state_dict(self, state_dict, **kwargs):
...
+        """
+        for param in self.encoder.parameters():
+            param.requires_grad = False
+
+        # Putting norm layers into eval mode stops running stats updates
+        for module in self.encoder.modules():
+            # _NormBase is the common base of _InstanceNorm and _BatchNorm classes
+            # These are the two classes that track running stats
+            if isinstance(module, torch.nn.modules.batchnorm._NormBase):
+                module.eval()
```
Comment (qubvel) [inline segmentation_models_pytorch/base/model.py:165]: We would also need a method to unfreeze the encoder, which will revert these ops, it would better to define one common method
```py
def _set_encoder_trainable(..

def freeze_encoder(...
    return self._set_encoder_trainable(False)
    
def unfreeze_encoder(...
    return self._set_encoder_trainable(True)
```

**Review by qubvel — COMMENTED** [review]
Hi @ricber, thanks for your contribution! Great work! I left some comments to improve the user experience. Let me know what you think!

File: segmentation_models_pytorch/base/model.py (hunk 3 above)
Comment (ricber) [reply segmentation_models_pytorch/base/model.py]: Addressed in 92b7e1c

**Comment by ricber** [comment]
Let me know if the docstrings are sufficient or if I should update the documentation in some way.

File: segmentation_models_pytorch/base/model.py [hunk 4]
```diff
@@ -138,28 +142,74 @@ def load_state_dict(self, state_dict, **kwargs):
...
-            # These are the two classes that track running stats
+            # _NormBase is the common base of classes like _InstanceNorm
+            # and _BatchNorm that track running stats
             if isinstance(module, torch.nn.modules.batchnorm._NormBase):
-                module.eval()
+                if mode:
+                    module.train()
+                else:
+                    # Putting norm layers into eval mode stops running stats updates
+                    module.eval()
```
Comment (qubvel) [inline segmentation_models_pytorch/base/model.py:188]:
```suggestion
                module.train(mode)
```

**Review by qubvel — COMMENTED** [review]
Thanks for the update, looks good to me! The last thing left is to add a few tests to make sure everything works and update docs 🤗

**Comment by ricber** [comment]
Are you OK with the tests in a699d67? Where should I update the docs?

**Comment by ricber** [comment]
I've addressed your last suggestions in 77fd490 and ffa3c5a

**Comment by ricber** [comment]
@qubvel just a little reminder of the open PR

**Comment by ricber** [comment]
Thank you for the merge 😉
